FileAlreadyExistsException when using multipartFormData body parser

#### Play version (2.6.2)
#### API Scala 2.11.7
#### OS MacOsX 10.10.5
#### JDK 1.8.0_141

I am using multipartFormData body parser to upload a file. When I run multiple upload requests at the same time I regularly get these two errors.
`FileAlreadyExistsException`
`DirectoryNotEmptyException`

Example:
` Execution exception[[FileAlreadyExistsException: /var/folders/1t/cwyz6701029fmqwqyy4_200r0000gn/T/playtemp3860699828886373735]]`
It seems that creation of temporary files is not thread safe.

### Stack trace for FileAlreadyExistsException:

! @7511lhil7 - Internal server error, for (POST) [/video/resumable/upload/j64y8tnk?resumableChunkNumber=52&resumableChunkSize=1048576&resumableCurrentChunkSize=1048576&resumableTotalSize=72968045&resumableType=video%2Fmp4&resumableIdentifier=72968045-f8fa54595d404450a487ce1f08012574m4v&resumableFilename=f8fa54595d404450a487ce1f08012574.m4v&resumableRelativePath=f8fa54595d404450a487ce1f08012574.m4v&resumableTotalChunks=69] ->
 
play.api.http.HttpErrorHandlerExceptions$$anon$1: Execution exception[[FileAlreadyExistsException: /var/folders/1t/cwyz6701029fmqwqyy4_200r0000gn/T/playtemp4016054458618813336]]
	at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:255)
	at play.api.http.DefaultHttpErrorHandler.onServerError(HttpErrorHandler.scala:180)
	at play.core.server.AkkaHttpServer$$anonfun$13$$anonfun$apply$1.applyOrElse(AkkaHttpServer.scala:252)
	at play.core.server.AkkaHttpServer$$anonfun$13$$anonfun$apply$1.applyOrElse(AkkaHttpServer.scala:251)
	at scala.concurrent.Future$$anonfun$recoverWith$1.apply(Future.scala:344)
	at scala.concurrent.Future$$anonfun$recoverWith$1.apply(Future.scala:343)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
	at play.api.libs.streams.Execution$trampoline$.execute(Execution.scala:70)
	at scala.concurrent.impl.CallbackRunnable.executeWithValue(Promise.scala:40)
	at scala.concurrent.impl.Promise$DefaultPromise.tryComplete(Promise.scala:248)
Caused by: java.nio.file.FileAlreadyExistsException: /var/folders/1t/cwyz6701029fmqwqyy4_200r0000gn/T/playtemp4016054458618813336
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:88)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
	at sun.nio.fs.UnixFileSystemProvider.createDirectory(UnixFileSystemProvider.java:384)
	at java.nio.file.Files.createDirectory(Files.java:674)
	at java.nio.file.Files.createAndCheckIsDirectory(Files.java:781)
	at java.nio.file.Files.createDirectories(Files.java:727)
	at play.api.libs.Files$DefaultTemporaryFileCreator.create(Files.scala:163)
	at play.core.parsers.Multipart$$anonfun$handleFilePartAsTemporaryFile$1.apply(Multipart.scala:121)
	at play.core.parsers.Multipart$$anonfun$handleFilePartAsTemporaryFile$1.apply(Multipart.scala:119)

### Stack trace for DirectoryNotEmptyException:

! @7511m7i2g - Internal server error, for (POST) [/video/resumable/upload/j64yewa1?resumableChunkNumber=9&resumableChunkSize=1048576&resumableCurrentChunkSize=1048576&resumableTotalSize=32203414&resumableType=video%2Fquicktime&resumableIdentifier=32203414-f854d2ed4e314292875ad2f2b7bcca96mov&resumableFilename=f854d2ed4e314292875ad2f2b7bcca96.mov&resumableRelativePath=f854d2ed4e314292875ad2f2b7bcca96.mov&resumableTotalChunks=30] ->
 
play.api.http.HttpErrorHandlerExceptions$$anon$1: Execution exception[[DirectoryNotEmptyException: /var/folders/1t/cwyz6701029fmqwqyy4_200r0000gn/T/playtemp5242047326278202622]]
	at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:255)
	at play.api.http.DefaultHttpErrorHandler.onServerError(HttpErrorHandler.scala:180)
	at play.core.server.AkkaHttpServer$$anonfun$13$$anonfun$apply$1.applyOrElse(AkkaHttpServer.scala:252)
	at play.core.server.AkkaHttpServer$$anonfun$13$$anonfun$apply$1.applyOrElse(AkkaHttpServer.scala:251)
	at scala.concurrent.Future$$anonfun$recoverWith$1.apply(Future.scala:344)
	at scala.concurrent.Future$$anonfun$recoverWith$1.apply(Future.scala:343)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
	at play.api.libs.streams.Execution$trampoline$.execute(Execution.scala:70)
	at scala.concurrent.impl.CallbackRunnable.executeWithValue(Promise.scala:40)
	at scala.concurrent.impl.Promise$DefaultPromise.tryComplete(Promise.scala:248)
Caused by: java.nio.file.DirectoryNotEmptyException: /var/folders/1t/cwyz6701029fmqwqyy4_200r0000gn/T/playtemp5242047326278202622
	at sun.nio.fs.UnixFileSystemProvider.implDelete(UnixFileSystemProvider.java:242)
	at sun.nio.fs.AbstractFileSystemProvider.deleteIfExists(AbstractFileSystemProvider.java:108)
	at java.nio.file.Files.deleteIfExists(Files.java:1165)
	at play.api.libs.Files$DefaultTemporaryFileCreator.play$api$libs$Files$DefaultTemporaryFileCreator$$playTempFolder(Files.scala:181)
	at play.api.libs.Files$DefaultTemporaryFileCreator.create(Files.scala:163)
	at play.core.parsers.Multipart$$anonfun$handleFilePartAsTemporaryFile$1.apply(Multipart.scala:121)
	at play.core.parsers.Multipart$$anonfun$handleFilePartAsTemporaryFile$1.apply(Multipart.scala:119)
	at play.core.parsers.Multipart$$anonfun$multipartParser$1$$anonfun$6.apply(Multipart.scala:79)
	at play.core.parsers.Multipart$$anonfun$multipartParser$1$$anonfun$6.apply(Multipart.scala:77)
	at akka.stream.impl.fusing.MapAsync$$anon$23.onPush(Ops.scala:1172)

--- a/framework/src/play/src/main/scala/play/api/libs/Files.scala
+++ b/framework/src/play/src/main/scala/play/api/libs/Files.scala
@@ -8,7 +8,6 @@ import java.lang.ref.Reference
 import java.nio.file.attribute.BasicFileAttributes
 import java.nio.file.{ Files => JFiles, _ }
 import java.time.{ Clock, Instant }
-import java.util.concurrent.atomic.AtomicReference
 import java.util.function.Predicate
 import javax.inject.{ Inject, Provider, Singleton }
 
@@ -19,7 +18,6 @@ import org.slf4j.LoggerFactory
 import play.api.Configuration
 import play.api.inject.ApplicationLifecycle
 
-import scala.annotation.tailrec
 import scala.concurrent.Future
 import scala.concurrent.duration._
 import scala.language.implicitConversions
@@ -149,38 +147,24 @@ object Files {
     private val logger = play.api.Logger(this.getClass)
     private val frq = new FinalizableReferenceQueue()
 
-    private val TempDirectoryPrefix = "playtemp"
-
     // Much of the PhantomReference implementation is taken from
     // the Google Guava documentation example
     //
     // https://google.github.io/guava/releases/19.0/api/docs/com/google/common/base/FinalizableReferenceQueue.html
     // Keeping references ensures that the FinalizablePhantomReference itself is not garbage-collected.
     private val references = Sets.newConcurrentHashSet[Reference[TemporaryFile]]()
-    private val _playTempFolder: AtomicReference[Option[Path]] = new AtomicReference(None)
 
-    override def create(prefix: String, suffix: String): TemporaryFile = {
-      JFiles.createDirectories(playTempFolder)
-      val tempFile = JFiles.createTempFile(playTempFolder, prefix, suffix)
-      createReference(new DefaultTemporaryFile(tempFile, this))
+    private val TempDirectoryPrefix = "playtemp"
+    private val _playTempFolder: Path = {
+      val tmpFolder = JFiles.createTempDirectory(TempDirectoryPrefix)
+      temporaryFileReaper.updateTempFolder(tmpFolder)
+      tmpFolder
     }
 
-    @tailrec private def playTempFolder: Path = {
-      val lastValue = _playTempFolder.get
-      val newFolder = lastValue match {
-        // We may need to recreate the file if it was deleted (e.g. by tmpwatch)
-        case Some(folder) if JFiles.exists(folder) =>
-          folder
-        case _ =>
-          JFiles.createTempDirectory(TempDirectoryPrefix)
-      }
-      if (_playTempFolder.compareAndSet(lastValue, Some(newFolder))) {
-        temporaryFileReaper.updateTempFolder(newFolder)
-        newFolder
-      } else {
-        JFiles.deleteIfExists(newFolder)
-        playTempFolder
-      }
+    override def create(prefix: String, suffix: String): TemporaryFile = {
+      JFiles.createDirectories(_playTempFolder)
+      val tempFile = JFiles.createTempFile(_playTempFolder, prefix, suffix)
+      createReference(new DefaultTemporaryFile(tempFile, this))
     }
 
     override def create(path: Path): TemporaryFile = {
@@ -192,7 +176,7 @@ object Files {
         override def finalizeReferent(): Unit = {
           references.remove(this)
           val path = tempFile.path
-          delete(tempFile)
+          deletePath(path)
         }
       }
       references.add(reference)
@@ -226,9 +210,9 @@ object Files {
      * Application stop hook which deletes the temporary folder recursively (including subfolders).
      */
     applicationLifecycle.addStopHook { () =>
-      Future.successful(JFiles.walkFileTree(playTempFolder, new SimpleFileVisitor[Path] {
+      Future.successful(JFiles.walkFileTree(_playTempFolder, new SimpleFileVisitor[Path] {
         override def visitFile(path: Path, attrs: BasicFileAttributes): FileVisitResult = {
-          logger.debug(s"stopHook: Removing leftover temporary file $path from $playTempFolder")
+          logger.debug(s"stopHook: Removing leftover temporary file $path from ${_playTempFolder}")
           deletePath(path)
           FileVisitResult.CONTINUE
         }

--- a/framework/src/play/src/test/scala/play/api/libs/TemporaryFileCreatorSpec.scala
+++ b/framework/src/play/src/test/scala/play/api/libs/TemporaryFileCreatorSpec.scala
@@ -6,6 +6,7 @@ package play.api.libs
 import java.io.File
 import java.nio.charset.Charset
 import java.nio.file.{ Path, Files => JFiles }
+import java.util.concurrent.{ CountDownLatch, ExecutorService, Executors }
 
 import org.specs2.mock.Mockito
 import org.specs2.mutable.{ After, Specification }
@@ -15,6 +16,10 @@ import play.api._
 import play.api.inject.DefaultApplicationLifecycle
 import play.api.libs.Files._
 import play.api.routing.Router
+
+import scala.concurrent.{ Await, ExecutionContext, Future }
+import scala.concurrent.duration._
+import scala.util.Try
 
 class TemporaryFileCreatorSpec extends Specification with Mockito {
 
@@ -39,6 +44,50 @@ class TemporaryFileCreatorSpec extends Specification with Mockito {
 
         parentDirectory.toFile.delete()
       }
+    }
+
+    "not have a race condition when creating temporary files" in {
+
+      // See issue https://github.com/playframework/playframework/issues/7700
+      // We were having problems by creating to many temporary folders and
+      // keeping track of them inside TemporaryFileCreator and between it and
+      // TemporaryFileReaper.
+
+      val threads = 25
+      val threadPool: ExecutorService = Executors.newFixedThreadPool(threads)
+
+      val lifecycle = new DefaultApplicationLifecycle
+      val reaper = mock[TemporaryFileReaper]
+      val creator = new DefaultTemporaryFileCreator(lifecycle, reaper)
+
+      try {
+        val executionContext = ExecutionContext.fromExecutorService(threadPool)
+
+        // Use a latch to stall the threads until they are all ready to go, then
+        // release them all at once. This maximizes the chance of a race condition
+        // being visible.
+        val raceLatch = new CountDownLatch(threads)
+
+        val futureResults: Seq[Future[Try[TemporaryFile]]] = for (_ <- 0 until threads) yield {
+          Future {
+            Try {
+              raceLatch.countDown()
+              creator.create("foo", "bar")
+            }
+          }(executionContext)
+        }
+
+        val results: Seq[Try[TemporaryFile]] = {
+          import ExecutionContext.Implicits.global // implicit for Future.sequence
+          Await.result(Future.sequence(futureResults), 30.seconds)
+        }
+
+        // If there is some failure, the test should fail then
+        results.exists(_.isFailure) must beFalse
+      } finally {
+        threadPool.shutdown()
+      }
+      ok
     }
 
     "recreate directory if it is deleted" in new WithScope() {

